### ISSUE TITLE
needless_borrow seems to be too eager

### ISSUE THREAD

**Issue by tspiteri** [rust-lang/rust-clippy#9774]
### Summary

Since beta 1.66, the `needless_borrow` lint is warning that "the borrowed expression implements the required traits" when passing a parameter by reference.

This can have an issue because let's say in the example below, the function `needs_to_reuse_string` will cause a version of `very_large_function` that takes `_s: &String` to be instantiated. Since `does_not_need_to_reuse_string` calls `very_large_function` with the same parameter type, no extra version is instantiated. If the clippy suggestion is obeyed, `does_not_need_to_reuse_string` will cause an extra instantiation of `very_large_function` which takes `_s: String`.

In summary, I don't get why the `needless_borrow` lint was broadened to warn about pass-by-reference when both pass-by-value and pass-by-reference are accepted. And even if it would be desirable to some, I don't agree with lumping that behaviour with `needless_borrow` which I find useful to catch stuff like `&already_a_reference`.

### Lint Name

needless_borrow

### Reproducer

I tried this code:

```rust
fn very_large_function<S: AsRef<[u8]>>(_s: S) {
    unimplemented!();
}

pub fn does_not_need_to_reuse_string() {
    let s = String::from("hello");
    very_large_function(&s);
}

pub fn needs_to_reuse_string() {
    let s = String::from("hello");
    very_large_function(&s);
    drop(s);
}
```

I saw this happen:

```
warning: the borrowed expression implements the required traits
 --> src/lib.rs:7:25
  |
7 |     very_large_function(&s);
  |                         ^^ help: change this to: `s`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow
  = note: `#[warn(clippy::needless_borrow)]` on by default
```

I expected to see no warnings.

### Version

```text
rustc 1.66.0-beta.1 (e080cc5a6 2022-11-01)
binary: rustc
commit-hash: e080cc5a659fb760c0bc561b722a790dad35b5e1
commit-date: 2022-11-01
host: x86_64-unknown-linux-gnu
release: 1.66.0-beta.1
LLVM version: 15.0.2
```

**Comment by tspiteri** [comment]
Just a note about the example I gave. I know ideally `very_large_function` would be written as
```rust
#[inline]
fn very_large_function<S: AsRef<[u8]>>(s: S) {
    very_large_function_common(s.as_ref());
}
fn very_large_function_common(_bytes: &[u8]) {
    unimplemented!()
}
```
My point was that the clippy suggestion can result in different generated code and thus the borrow affects generated code unlike passing `&&&&&&5` to a function that takes `&i32` like the lint documentation example.
